src/psyclone/dynamo0p3.py

```diff
@@ -2683,19 +2767,66 @@ def _invoke_declarations(self, parent):
                        routine to which to add declarations.
         :type parent: :py:class:`psyclone.f2pygen.SubroutineGen`
 
+        :raises InternalError: for an unsupported intrinsic type of field \
+                               argument data (other than "real" or "integer").
+        :raises GenerationError: if the same field has different data \
+                                 types in different kernel calls within \
+                                 the same Invoke.
+
         '''
         # Add the Invoke subroutine argument declarations for fields
-        fld_args = self._invoke.unique_declarations(
-            argument_types=LFRicArgDescriptor.VALID_FIELD_NAMES)
-        # Create a list of field names
-        fld_arg_list = [arg.declaration_name for arg in fld_args]
-        if fld_arg_list:
+        fld_args = []
+        for kern in self._invoke.schedule.kernels():
+            fld_args.extend(psyGen.args_filter(
+                kern.arguments.args,
+                arg_types=LFRicArgDescriptor.VALID_FIELD_NAMES))
+        # Create lists of field names for real- and integer-valued fields
+        real_fld_arg_list = []
```

Comment:
I tried that originally and it only made it cosmetically better. The proper solution would be to modify `unique_declarations` and `unique_declns_by_intent` in `psyGen.py` to filter kernel arguments by their `intrinsic_type` (optional parameter) in addition to the `argument_type`, similar to how I did it here for the `unique_proxy_declarations`. This would return lists of unique names for `real` and `integer` fields whilst still making possible that the same field name is in both lists if the user made an error when calling different kernels within the same Invoke.

In fact, this would be my preferred solution, as it would also significantly improve the implementation of `DynScalarArg`. There I also filtered scalar by intrinsic type with `if` statements after they were returned from `unique_declns_by_intent`. Also, currently there is no check that the same scalar name can be declared as both `real` and `integer` when calling different kernels within the same Invoke. Without modifying `unique_declarations` and `unique_declns_by_intent` that check will be equally clumsy as in `DynFields`.

Reply:
Great.